### DISCUSSION TITLE
Suggestion: Movement overhaul (Parkour)

### DISCUSSION THREAD

**Discussion by Kegdo-Jr** [cataclysmbn/Cataclysm-BN#2882] [General]
### Is your feature request related to a problem? Please describe.

Let's say you are being chased by a hulk on an apartment tower's roof. There is a roof just two tiles away, but since it's two and not one, you can't use the hardcoded jump to escape certain death. 

In general, there is absolutely no roof-to-roof traversal except a hardcoded cata leftover for 1-tile gaps, nor any point in going lightweight in late-game, since armor is 90% of the time better than anything else since you simply cannot outrun things.

### Describe the solution you'd like

Running overhaul:
When running, instead of gaining a flat increase to tile travel speed, for each move done in the same direction, you get a slight increase to 'Momentum', - a new var that shows your potential kinetic energy, - that increases your move speed for up to twice as much of your default speed. The Momentum is capped, but the limit can be increased with more Dexterity.
The more Momentum you have, the farther you can [Q] Leap (ending the game forever shouldn't be a keybind in the first place), jumping over up to 5 (or 4) passable tiles. This Leap puts you in a controllable jump state, which allows you to do cool schtuff. In this state, you can't move in a preferred direction, instead being led by the force of physics to your untimely demise. You can press 5/. (pass one turn) to get one tile closer to your destination. This turn has a fixed time of 0.25 seconds. Leaping uses a lot of your stamina, roughly five percent or so.
If you meet with an impassable object or a wall during a Leap, you smash into it, taking damage and going into an uncontrollable fall.

PARKOUR!!1:
New interaction skill - parkour. Increased by doing the following:
 
Roll - If you land on a lower Z-level during a controllable jump, instead of taking flat random damage, you accumulate it and reduce the total by `( X * Z * O * (M:2) ):P`, where X - damage, Z - amount of Z-levels passed, O - obstacle softness (if you fall on a zombie, for example, this would be 0.7, if you land on a couch - 0.2, a landing pad - 0, and a fridge - 1.3), M - momentum (you gain momentum for each Z-level passed, which is good for things listed below, but bad for your legs and spine) and P - parkour skill. If your parkour skill is 0, instead of dividing damage you multiply it by 1.1. The proc message is "You roll, dissipating the most of the impact's force!" or "You crash on a landing pad, negating the fall's force!". Yes, Dying Light reference. No, you can't use a car's roof as a landing pad.

TicTac/Wallrun - During a controllable jump, if you try to move into a wall with enough Momentum, you start running on it in the chosen direction that is close to the Leap's initial one (so a Leap in NORTH (8) direction, would allow you to wallrun in NORTH-WEST, NORTH and NORTH-EAST directions (7, 8 and 9 on numpad)), spending Momentum for each tile crossed. You can Leap while running on a wall to get a bit of Momentum back, therefore making it possible to cross 6+ tile long gaps, provided the gap is narrow enough and you are fast.

Wall climb - During a controllable jump, you can go up (<, like with the stairs) near a wall, spending a portion of your Momentum. It should be possible to go up multiple walls that way, but the limit is 3 with 19-20 Dexterity worth of maximum Momentum. After the climb, you are put in the air, so you can either move to a nearby passable tile, like a roof, wall climb again or Leap in a picked direction, making it possible to actually climb things.

Vault - with enough Momentum, you can automatically vault over passable tiles that slow you down (tables, bars, chairs, etc.), spending some in the proccess. Jump state not required.

Cat leap - During a controllable jump, if you move into a passable tile while being in the air (empty space that would usually send you flying to the lower Z-level), you can grab the ledge and pull yourself to it, but only if you have enough strength to do so; 6 STR for 1 Z-level, 10 for 2, 14 for 3 and so on. This negates your fall damage.

The higher is your parkour skill, the more Momentum you can have, the higher is your sprint speed limit, and the less Momentum is required to Leap.

To show when you can Leap, the moving intent indicator (C, W, R) should change its color from gray to red when you start running and your Momentum is too low, and from red to green when it's enough. 

With this change, Parkour master should also always gain 30% more Momentum and Momentum cap, as well as negate more fall damage using a Roll. Bad knees is the opposite.

Parkour-related constructions:
Landing pad - made out of crazy amounts of pillows, three matresses and two long ropes, this thing would completely negate any fall damage if you land on it. It's nearly impossible to land on ONE tile of safety after falling 3+ Z-levels down, so if someone actually manages it, it's completely fine since they most likely spent a lot of time looting twenty houses for pillows and matresses to make a 9x9 square of beddings.

Zipline - basically a thing that sends  you from one point to another, but only if the starting point is above the ending one. It puts you in a controllable jump state, so you can Leap or get shot at any moment.

### Additional context

i hate the hardcoded 1-tile jump so much that i wrote this entire thing.

also wings would still negate your fall damage and outright ignore the Roll technique.

**Comment by Empyrean-Heaven** [comment]
Sounds fun! Not sure how to code it in, but I like it!
